k-distribution/src/main/scripts/lib/pyk/kast.py

```diff
@@ -1,255 +1,1065 @@
+
+    # TODO consider nullable fields and make sure their value can be erased
+    def copy(self, *, regex: Optional[str] = None, precede_regex: Optional[str] = None, follow_regex: Optional[str] = None) -> 'KRegexTerminal':
+        regex = regex if regex is not None else self.regex
+        precede_regex = precede_regex if precede_regex is not None else self.precede_regex
+        follow_regex = follow_regex if follow_regex is not None else self.follow_regex
+        return KRegexTerminal(regex=regex, precede_regex=precede_regex, follow_regex=follow_regex)
+
+
+@final
+@dataclass(frozen=True)
+class KNonTerminal(KProductionItem):
+    sort: KSort
+
+    def __init__(self, sort: KSort):
+        object.__setattr__(self, 'sort', sort)
+
+    @classmethod
+    def from_dict(cls: Type['KNonTerminal'], d: Dict[str, Any]) -> 'KNonTerminal':
+        cls._check_node(d)
+        return KNonTerminal(sort=KSort.from_dict(d['sort']))
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {'node': 'KNonTerminal', 'sort': self.sort.to_dict()}
+
+    def copy(self, *, sort: Optional[KSort] = None) -> 'KNonTerminal':
+        sort = sort or self.sort
+        return KNonTerminal(sort=sort)
+
+
+@final
+@dataclass(frozen=True)
+class KProduction(KSentence):
+    sort: KSort
+    items: Tuple[KProductionItem, ...]
+    klabel: str
+    att: KAtt
+
+    def __init__(self, sort: KSort, items: Iterable[KProductionItem] = (), klabel='', att=EMPTY_ATT):
+        object.__setattr__(self, 'sort', sort)
+        object.__setattr__(self, 'items', tuple(items))
+        object.__setattr__(self, 'klabel', klabel)
+        object.__setattr__(self, 'att', att)
+
+    @property
+    def arity(self) -> int:
+        return len(self.items)
+
+    @classmethod
+    def from_dict(cls: Type['KProduction'], d: Dict[str, Any]) -> 'KProduction':
+        cls._check_node(d)
+        return KProduction(
+            sort=KSort.from_dict(d['sort']),
+            items=(KProductionItem.from_dict(item) for item in d['productionItems']),
+            klabel=d.get('klabel', ''),
+            att=KAtt.from_dict(d['att']) if d.get('att') else EMPTY_ATT,
+        )
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {
+            'node': 'KProduction',
+            'sort': self.sort.to_dict(),
+            'productionItems': [item.to_dict() for item in self.items],
+            'klabel': self.klabel or None,
+            'att': self.att.to_dict(),
+        }
+
+    def copy(
+        self,
+        *,
+        sort: Optional[KSort] = None,
+        items: Optional[Iterable[KProductionItem]] = None,
+        klabel: Optional[str] = None,
+        att: Optional[KAtt] = None,
+    ) -> 'KProduction':
+        sort = sort or self.sort
+        items = items if items is not None else self.items
+        klabel = klabel if klabel is not None else self.klabel
+        att = att or self.att
```

Comment:
Why do some use the `sort = sort or self.sort`, and other the `sort = sort if sort is not None else self.sort` technique?

Reply:
For `sort: Optional[KSort]` the only falsy value is `None`, but for `sort: Optional[str]` the falsy values are `''` and `None`. I still need to add precondition checks to make sure that `sort: str` (and all similar `str` fields) are nonempty, and then the expressions can be uniformly simplified to `sort or self.sort`.